Source/ablastr/fields/IntegratedGreenFunctionSolver.H

```diff
@@ -0,0 +1,225 @@
+/* Copyright 2019-2022 Axel Huebl, Remi Lehe
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef ABLASTR_IGF_SOLVER_H
+#define ABLASTR_IGF_SOLVER_H
+
+#include <ablastr/constant.H>
+#include <ablastr/utils/Communication.H>
+#include <ablastr/utils/TextMsg.H>
+#include <ablastr/warn_manager/WarnManager.H>
+#include <ablastr/math/fft/AnyFFT.H>
+
+#include <AMReX_Array.H>
+#include <AMReX_Array4.H>
+#include <AMReX_BLassert.H>
+#include <AMReX_Box.H>
+#include <AMReX_BoxArray.H>
+#include <AMReX_Config.H>
+#include <AMReX_DistributionMapping.H>
+#include <AMReX_FArrayBox.H>
+#include <AMReX_FabArray.H>
+#include <AMReX_Geometry.H>
+#include <AMReX_GpuControl.H>
+#include <AMReX_GpuLaunch.H>
+#include <AMReX_GpuQualifiers.H>
+#include <AMReX_IndexType.H>
+#include <AMReX_IntVect.H>
+#include <AMReX_LO_BCTYPES.H>
+#include <AMReX_MFIter.H>
+#include <AMReX_MFInterp_C.H>
+#include <AMReX_MLLinOp.H>
+#include <AMReX_MultiFab.H>
+#include <AMReX_Parser.H>
+#include <AMReX_REAL.H>
+#include <AMReX_SPACE.H>
+#include <AMReX_Vector.H>
+
+
+#include <array>
+#include <optional>
+
+
+namespace {
+
+using namespace amrex::literals;
+
+// Implements equation 2 in https://journals.aps.org/prab/pdf/10.1103/PhysRevSTAB.10.129901 with some modification to symmetrize the function
+AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
+amrex::Real IntegratedPotential( amrex::Real x, amrex::Real y, amrex::Real z ) {
+    amrex::Real const r = std::sqrt( x*x + y*y + z*z );
+    amrex::Real const G =
+        - 0.5_rt * z*z * std::atan( x*y/(z*r) )
+        - 0.5_rt * y*y * std::atan( x*z/(y*r) )
+        - 0.5_rt * x*x * std::atan( y*z/(x*r) )
+        + y*z*std::asinh( x/std::sqrt(y*y + z*z) )
+        + x*z*std::asinh( y/std::sqrt(x*x + z*z) )
+        + x*y*std::asinh( z/std::sqrt(x*x + y*y) );
+    return G;
+}
+
+inline void
+computePhiIGF ( amrex::MultiFab const & rho,
+                amrex::MultiFab & phi,
+                std::array<amrex::Real, 3> const cell_size,
+                amrex::BoxArray const ba ) {
+
```

Comment:
Looking at how this is used, I do not think it benefits from `inline`/inlining, does it?

I would propose, in the sense of improving compile time, to move the body `{}` of this function into a `.cpp` file.

Reply:
ping @aeriforme

Reply:
`inline` was requested by Clang-Tidy. 
Should I create a `.cpp` for the body of this function only then? 
Wouldn't it be less readable?

Reply:
Yes, please move it to a `.cpp`. This is about compile-time - if we unnecessarily put implementations in `.H` files then our compilation times for WarpX and ImpactX will go up over time.